stencila/pyla/interpreter.py

```diff
@@ -227,13 +228,115 @@ def traverse_list(self, _list: typing.List, compilation_result: DocumentCompilat
 class Interpreter:
     """Execute a list of code blocks, maintaining its own `globals` scope for this execution run."""
 
+    """
+    List of values for the `programmingLanguage` property that are handled
+    by this interpreter.
+    """
+    PROGRAMMING_LANGUAGES = ['py', 'python']
+
+    """
+    JSON Schema specification of the types of nodes that this intertpreter's
+    `compile` and `execute` methods are capable of handling
+    """
+    CODE_CAPABILITIES = {
+        'type': 'object',
+        'required': ['node'],
+        'properties': {
+            'node': {
+                'type': 'object',
+                'required': ['type', 'programmingLanguage'],
+                'properties': {
+                    'type': {
+                        'enum': ['CodeChunk', 'CodeExpression']
+                    },
+                    'programmingLanguage': {
+                        'enum': PROGRAMMING_LANGUAGES
+                    }
+                }
+            }
+        }
+    }
+
+    """
+    The manifest of this interpreter's capabilities and addresses.
+
+    Conforms to Executa's
+    [Manifest](https://github.com/stencila/executa/blob/v1.4.0/src/base/Executor.ts#L63)
+    interface.
+    """
+    MANIFEST = {
+        'version': 1,
+        'capabilities': {
+            'compile': CODE_CAPABILITIES,
+            'execute': CODE_CAPABILITIES
+        },
+        'addresses': {
+            'stdio': {
+                'type': 'stdio',
+                'command': sys.executable,
+                'args': ['-m', 'stencila.pyla', 'serve']
+            }
+        }
+    }
+
     globals: typing.Dict[str, typing.Any]
     locals: typing.Dict[str, typing.Any]
 
     def __init__(self) -> None:
         self.globals = {}
         self.locals = {}
 
+    @staticmethod
+    def compile(node: Node) -> Node:
+        """Compile a `CodeChunk`"""
+        if isinstance(node, CodeChunk) and Interpreter.is_python_code(node):
```

Comment:
This should work in the same way as the `handle_code` method on `Interpreter`. The `if` statements aren't necessary, and the `set_code_imports` method should be used (this checks for doubles ups and stops adding imports if a `.` is found).

Maybe the code for parsing in `handle_code` could be moved into this `compile` method and `handle_code` could call this instead?

Reply:
Thanks, I didn't see `handle_code`.

It feels like `set_code_imports` would be better to be in `CodeChunkParser` i.e `cc_result.imports` is already normalised. What do you think?

As for the `if` statements, without those, empty array attributes get returned on the code chunk, which while not a problem, is unnecessary and a bit noisy when debugging.

> Maybe the code for parsing in handle_code could be moved into this compile method and handle_code could call this instead?

Yep, would be good to avoid any duplication.

Reply:
> It feels like `set_code_imports` would be better to be in `CodeChunkParser` i.e `cc_result.imports` is already normalised. What do you think?

That would be good, it could probably be refactored so this is the case. Let me know if you want me to take over this. I would probably turn these into `@property` methods that would return `None` instead of empty arrays so that everything is cleaner. `imports` could do a JIT normalisation of the imports array.

Reply:
I think we should hold off on any refactoring for now and do a more comprehensive refactoring later taking into account the new functionality in Executa (that might make things like `class DocumentCompiler` redundant) and experiences from developing interpreters for other languages e.g. `bash` and `R`.

If that's OK with you, can we go ahead and merge this?

Reply:
@nokome I am done with refactoring and this can now be merged.

Reply:
Thanks